### PR TITLE
Add gRPC `RaftService`

### LINKED ISSUE #483: Consensus: Introduce internal gRPC service for Raft messages
Introduce RaftMessageService here `.add_service(RaftMessageService::new(...))`
https://github.com/qdrant/qdrant/blob/bd8ad015c8a99597b95efbb0897a4bb6b795aa7f/src/tonic/mod.rs#L98

### PR DESCRIPTION
**Pull request by e-ivkov**
Closes #483

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -39,12 +39,12 @@ tokio = { version = "~1.17", features = ["full"] }
...
 num-traits = "0.2.14"
-prost = "0.10.1"
-prost-types = "0.10.1"
 
-raft = { version = "0.6.0", features = ["prost-codec"], default-features = false, optional = true }
+# Consensus related crates
+raft = { version = "=0.6.0", features = ["prost-codec"], default-features = false, optional = true }
 slog =  { version = "2.7.0", optional = true }
 slog-stdlog = { version = "4.1.1", optional = true }
+prost = { version = "=0.7.0", optional = true }
```
Comment (agourlay) [inline Cargo.toml:47]: did it fail with `prost` 0.10.1?

File: src/tonic/api/raft_api.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+impl Raft for RaftService {
+    async fn send(&self, mut request: Request<RaftMessageBytes>) -> Result<Response<()>, Status> {
+        let message = <RaftMessage as prost::Message>::decode(&request.get_mut().message[..])
+            .map_err(|err| {
+                Status::invalid_argument(format!("Failed to parse raft message: {err}"))
+            })?;
+        self.0
+            .lock()
+            .map_err(|_| Status::internal("Can't capture the Raft message sender lock"))?
+            .send(consensus::Message::FromPeer(Box::new(message)))
```
Comment (agourlay) [inline src/tonic/api/raft_api.rs:25]: random idea: does it make sense to use a channel with a given capacity to be able to apply some backpressure at the Raft level?

Comment (agourlay) [inline src/tonic/api/raft_api.rs:25]: random idea: does it make sense to use a channel with a given capacity to be able to apply some backpressure at the Raft level?

**Review by generall — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (e-ivkov) [reply Cargo.toml:47]: Yes, `raft-rs` uses prost `0.7.0` so basically there is no other way to serialize raft messages

**Review by agourlay — APPROVED** [review]

File: src/tonic/api/raft_api.rs (hunk 2 above)
Comment (e-ivkov) [reply src/tonic/api/raft_api.rs:25]: https://github.com/qdrant/qdrant/issues/495

**Pull request merged**